content/learning/operator/operator.md

```diff
@@ -65,10 +65,100 @@ Kubebuilder will generate the controller code only after generating CRDs, so eve
+An example is a resource that needs reconciling only if a specific label is modified.
+This operation can be performed thanks to the [Predicates](https://sdk.operatorframework.io/docs/building-operators/golang/references/event-filtering/).
+In order to use them you should modify the setup function explained above in the following way:
+```
+func (r *ObjectToReconcile) SetupWithManager(mgr ctrl.Manager) error {
+    klog.Info("setup manager")
+    return ctrl.NewControllerManagedBy(mgr).
+	For(&principal_resource, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+        Owns(&second_resource_to_watch, builder.WithPredicates(someExamplePredicate())).
+        Complete(r)
+}
+```
+The SetupWithManager function is different from the above one since now, together with the watched resource, there is an optional parameter (the _predicate_) that specifies when the reconcile is triggered.
+
+```
+func someExamplePredicate() predicate.Predicate {
+    return predicate.Funcs{
+        CreateFunc: func(e event.UpdateEvent) bool {
+            // Here you can compare old version and new version of the watched object
+            return condition_to_evaluate
+        },
+        UpdateFunc: func(e event.UpdateEvent) bool {
+            // Here you can compare old version and new version of the watched object
+            return condition_to_evaluate
+        },
+        DeleteFunc: func(e event.DeleteEvent) bool {
+            // Here you can compare old version and new version of the watched object
+            return condition_to_evaluate
+        },
+    }
+}
+```
+
+A Predicate takes an event of a particular type (Create / Update / Delete) and returns true if the event should be processed by reconcile.
+
+3. Another problem that you can face relates to the creation and the update of resources in the cluster.
+Given that a controller must be _idempotent_ (superfluous reconcile has no effect, e.g. resources already present in the cluster must be updated and not recreated every reconcile), it is better to use [CreateOrUpdate](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.0/pkg/controller/controllerutil#CreateOrUpdate), a function that is intrinsically idempotent and it is part of the `controller-utils` package.
+In order to use it you have to create an initial resource composed only by ObjectMeta's data; then you have to pass a _mutate function_ as a parameter, which represents a function that works on the resource itself for example adding labels, modifying annotations and describing specifications.
+
+Here there is a CreateOrUpdate example for a VirtualMachine:
+```
+vm := v1.VirtualMachine{ObjectMeta: metav1.ObjectMeta{Name: "name", Namespace: "namespace"}}
+    _, err = ctrl.CreateOrUpdate(ctx, r.Client, &vm, func() error {
+        UpdateVirtualMachineSpec(&vm)
+        UpdateLabels(vm.Spec.Template.ObjectMeta.Labels)
+        return ctrl.SetControllerReference(resource, &vm, r.Scheme)
+	})
```

Comment:
First, check the indentation. Second, I would stick to an example involving core Kubernetes components (e.g. deployment, service, secret, configmap, ...), as VMs are custom resources which may not be well known and may raise confusion.